src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php

```diff
@@ -297,23 +297,30 @@ protected function stream(string $view, array $parameters = [], StreamedResponse
      * * if the form is submitted but invalid, $render is called and a 422 HTTP status code is set if the current status hasn't been customized
      * * if the form is submitted and valid, $onSuccess is called, usually this method saves the data and returns a 303 HTTP redirection
      *
-     * @param callable(FormInterface, mixed): Response $onSuccess
-     * @param callable(FormInterface, mixed): Response $render
+     * For both callables, instead of "mixed", you can use your form's data class as a type-hint for argument #2.
+     *
+     * @param callable(FormInterface, mixed, Request): Response $onSuccess
+     * @param callable(FormInterface, mixed, Request): Response $render
      */
     public function handleForm(FormInterface $form, Request $request, callable $onSuccess, callable $render): Response
     {
         $form->handleRequest($request);
 
         $submitted = $form->isSubmitted();
-
         $data = $form->getData();
-        if ($submitted && $form->isValid()) {
-            return $onSuccess($form, $data);
+
+        if ($isValid = $submitted && $form->isValid()) {
+            $response = $onSuccess($form, $data, $request);
+        } else {
+            $response = $render($form, $data, $request);
+
+            if ($submitted && 200 === $response->getStatusCode()) {
+                $response->setStatusCode(Response::HTTP_UNPROCESSABLE_ENTITY);
+            }
         }
 
-        $response = $render($form, $data);
-        if ($submitted && 200 === $response->getStatusCode()) {
-            $response->setStatusCode(Response::HTTP_UNPROCESSABLE_ENTITY);
+        if (!$response instanceof Response) {
```

Comment:
Wouldn't this break using `@Template` and other kind of view listeners? Couldn't we just check that the return value isn't null?